logging/resource.go

```diff
@@ -34,165 +33,172 @@ type commonResource struct{ *mrpb.MonitoredResource }
 
 func (r commonResource) set(l *Logger) { l.commonResource = r.MonitoredResource }
 
-var detectedResource struct {
-	pb   *mrpb.MonitoredResource
-	once sync.Once
+var detectedResource = struct {
+	pb    *mrpb.MonitoredResource
+	attrs internal.ResourceAtttributesGetter
+	once  *sync.Once
+}{
+	attrs: internal.ResourceAttributes(),
+	once:  new(sync.Once),
+}
+
+func metadataProjectID() (string, bool) {
+	return detectedResource.attrs.Metadata("project/project-id")
+}
+
+func metadataZone() (string, bool) {
+	zone, ok := detectedResource.attrs.Metadata("instance/zone")
+	if ok {
+		return zone[strings.LastIndex(zone, "/")+1:], true
+	}
+	return "", false
+}
+
+func metadataRegion() (string, bool) {
+	region, ok := detectedResource.attrs.Metadata("instance/region")
+	if ok {
+		return region[strings.LastIndex(region, "/")+1:], true
+	}
+	return "", false
 }
 
 // isAppEngine returns true for both standard and flex
 func isAppEngine() bool {
-	_, service := os.LookupEnv("GAE_SERVICE")
-	_, version := os.LookupEnv("GAE_VERSION")
-	_, instance := os.LookupEnv("GAE_INSTANCE")
-
-	return service && version && instance
+	_, serviceOK := detectedResource.attrs.EnvVar("GAE_SERVICE")
+	_, versionOK := detectedResource.attrs.EnvVar("GAE_VERSION")
+	_, instanceOK := detectedResource.attrs.EnvVar("GAE_INSTANCE")
+	return serviceOK && versionOK && instanceOK
 }
 
 func detectAppEngineResource() *mrpb.MonitoredResource {
-	projectID, err := metadata.ProjectID()
-	if err != nil {
+	projectID, ok := metadataProjectID()
+	if !ok {
```

Comment:
Correct me if I wrong, but failed metadata reading will make code to bail out, right? If thats a case, why not to use alternate way to retrieve projectID like `os.Getenv("GOOGLE_CLOUD_PROJECT")`?

Reply:
Talked offline